docs/tasks/administer-cluster/dns-custom-nameservers.md

```diff
@@ -55,66 +39,82 @@ data:
-kube-dns service. Queries for domains rooted in the configured cluster domain
-suffix (any address ending in “.cluster.local” in the example above) are
-answered by the kube-dns service. All other queries, such as
-www.kubernetes.io, are forwarded to the upstream nameserver inherited from
-the node.
+### "ClusterFirst" DNS Policy
+
+If the `dnsPolicy` is set to "ClusterFirst", name resolution is handled differently, *depending on whether stub-domain and upstream DNS servers are configured*.
+
+If they are *not* configured, any query that does not match the configured cluster domain suffix, such as "www.kubernetes.io", is forwarded to the upstream nameserver inherited from the node.
+
+If stub domains and upstream DNS servers are configured (as in the [previous example](#configuring-stub-domain-and-upstream-dns-servers)), DNS queries will be
+routed according to the following flow:
+1. The query is first sent to the DNS caching layer in kube-dns.
+
+1. From the DNS caching layer, the suffix of the request is examined and then forwarded to the appropriate DNS, based on the following cases:
+   * *Names with the cluster suffix* (e.g.".cluster.local"): The request is sent to kube-dns.
+
+   * *Names with the stub domain suffix* (e.g. ".acme.local"): The request is sent to the configured custom DNS resolver (e.g. listening at 1.2.3.4).
+
+   * *Names without a matching suffix* (e.g."widget.com"): The request is forwarded to the upstream DNS (e.g. Google public DNS servers at 8.8.8.8 and 8.8.4.4).
+
+![DNS lookup flow](/docs/tasks/administer-cluster/dns-custom-nameservers/dns.png)
 
 ### ConfigMap options
 
 Options for the kube-dns `kube-system:kube-dns` ConfigMap
 
 | Field | Format | Description |
 | ----- | ------ | ----------- |
-| stubDomains (optional) | A JSON map using a DNS suffix key (e.g. “acme.local”) and a value consisting of a JSON array of DNS IPs. | The target nameserver may itself be a Kubernetes service. For instance, you can run your own copy of dnsmasq to export custom DNS names into the ClusterDNS namespace. |
-| upstreamNameservers (optional) | A JSON array of DNS IPs. | Note: If specified, then the values specified replace the nameservers taken by default from the node’s /etc/resolv.conf Limits: a maximum of three upstream nameservers can be specified. |
+| `stubDomains` (optional) | A JSON map using a DNS suffix key (e.g. “acme.local”) and a value consisting of a JSON array of DNS IPs. | The target nameserver may itself be a Kubernetes service. For instance, you can run your own copy of dnsmasq to export custom DNS names into the ClusterDNS namespace. |
+| `upstreamNameservers` (optional) | A JSON array of DNS IPs. | Note: If specified, then the values specified replace the nameservers taken by default from the node’s `/etc/resolv.conf`. Limits: a maximum of three upstream nameservers can be specified. |
+
+{% endcapture %}
+
+{% capture whatsnext %}
 
-### Additional examples
+## Additional examples
```

Comment:
There's a *What's next* h2 header automatically generated, which looks empty if this is also an h2. I think either these additional examples should be h3s under *What's next* or they should be moved up into the *steps* section.